### PR TITLE
Adding a note about the Android Studio Flamingo gradle mismatch

### PR DESCRIPTION
**Pull request by sfshaza2**
This PR supercedes PR https://github.com/flutter/website/pull/8490.

Staged here:
main page:  https://parlough-android-flutter.web.app/
migration page:  https://parlough-android-flutter.web.app/release/breaking-changes/android-java-gradle-migration-guide
the "go" link:  https://parlough-android-flutter.web.app/go/android-java-gradle-error

(This is the staged go link. The actual go link will be docs.flutter.dev rather than `parlough-android-flutter.web.app`.)

@reidbaker @camsim99 @mariamhas @ericwindmill @RedBrogdon  @timsneath

### CODE REVIEW

**Review by domesticmouse — APPROVED** [review]

**Comment by camsim99** [comment]
@sfshaza2 Was the go link go/android-android-gradle-error intended to be go/android-java-gradle-error?

File: firebase.json [hunk 1]
```diff
@@ -175,6 +175,7 @@
       { "source": "/go/android-assets-in-flutter", "destination": "https://docs.google.com/document/d/1jymgQYxRTe5rdprt74ERh7Jsa0lfnRuMnkOmJusLWsE/edit", "type": 301 },
       { "source": "/go/android-embedding-dependencies", "destination": "https://docs.google.com/document/d/1vITp2mUZRa-cmll0sPH0zjNgPlyvOMx7awxPNRAPyic/edit", "type": 301 },
       { "source": "/go/android-embedding-move", "destination": "https://docs.google.com/document/d/1nQujwZfEe3QOHTyZn160eImpoJOYioADP09DtumcLcc/edit?ts=5d8041db#", "type": 301 },
+      { "source": "/go/android-android-gradle-error", "destination": "/release/breaking-changes/android-java-gradle-migration-guide", "type": 301 },
```
Comment (parlough) [inline firebase.json:178]:
```suggestion
      { "source": "/go/android-java-gradle-error", "destination": "/release/breaking-changes/android-java-gradle-migration-guide", "type": 301 },
```

File: src/release/breaking-changes/android-java-gradle-migration-guide.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+![Workflow to upgrade Gradle]({{site.url}}/assets/images/docs/releaseguide/android-studio-flamingo-gradle-upgrade.png){:width="85%"}
+
+## Solution #2: Manual fix at the command line
+
+Do the following from the top of your Flutter project.
+
+**Step 1**: Go to the gradle directory for your project:
+
+```sh
+cd android/gradle/wrapper
```
Comment (parlough) [inline src/release/breaking-changes/android-java-gradle-migration-guide.md:69]:
````suggestion
```terminal
$ cd android/gradle/wrapper
````

File: src/release/breaking-changes/android-java-gradle-migration-guide.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+```
+
+**Step 2**: Edit the `gradle-wrapper.properties` file to
+  change the `distributionUrl` field to the new Gradle version:
+
+```sh
```
Comment (parlough) [inline src/release/breaking-changes/android-java-gradle-migration-guide.md:75]:
````suggestion
```properties
````

File: src/_includes/banner.html [hunk 4]
```diff
@@ -3,11 +3,7 @@
 {% endcomment -%}
 <div class="site-banner site-banner--default" role="alert">
   Flutter is back at Google I/O on May 10th!
```
Comment (timsneath) [inline src/_includes/banner.html:5]: Can we have just one banner? It looks a bit confusing to both be promoting I/O *and* this change...

File: src/release/breaking-changes/android-java-gradle-migration-guide.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.0.2-all.zip
```
Comment (timsneath) [inline src/release/breaking-changes/android-java-gradle-migration-guide.md:76]: Is the backslash an error here? `https\://`

File: src/release/breaking-changes/android-java-gradle-migration-guide.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    command reports which version of Java is used.
+* If you upgrade Gradle to a release _newer_ than 7.6,
+  you might (though it's unlikely) encounter issues
+  that result from changes to Gradle, such as
+  [deprecated Gradle classes][], or changes to the
+  Android file structure, such as
+  [splitting out ApplicationId from PackageName][].
+  If this occurs, downgrade to the 7.6 release of
+  Gradle.
+* If, as part of this fix, you upgraded to Flutter 3.10
```
Comment (timsneath) [inline src/release/breaking-changes/android-java-gradle-migration-guide.md:109]: This bullet seems completely orthogonal, no? Upgrading to Flutter 3.10 is not connected to upgrading Android Studio.

File: src/release/breaking-changes/android-java-gradle-migration-guide.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  	[…
+  	 …
+  	 … 209 more lines of Groovy and Gradle stack trace …
+  	 …
+  	 …]
+  	at java.base/java.lang.Thread.run(Thread.java:833)
+```
+
+This error occurs because Android Studio Flamingo
+updates the Java SDK from 11 to 17, which
```
Comment (timsneath) [inline src/release/breaking-changes/android-java-gradle-migration-guide.md:39]: updates the _installed_ Java SDK, perhaps?

File: src/release/breaking-changes/android-java-gradle-migration-guide.md [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+isn't compatible with the version of the Java SDK
+used to build the app's Gradle file when the project
+was originally created with `flutter create`.
```
Comment (timsneath) [inline src/release/breaking-changes/android-java-gradle-migration-guide.md:42]: Long run-on sentence?

File: src/release/breaking-changes/android-java-gradle-migration-guide.md (hunk 5 above)
Comment (parlough) [reply src/release/breaking-changes/android-java-gradle-migration-guide.md:76]: The backslash is [expected](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Properties.html#store(java.io.Writer,java.lang.String)) in the `.properties` file: https://docs.gradle.org/current/userguide/gradle_wrapper.html#customizing_wrapper

File: src/release/breaking-changes/android-java-gradle-migration-guide.md (hunk 2 above)
Comment (parlough) [reply src/release/breaking-changes/android-java-gradle-migration-guide.md:69]: Yeah, we've generally used this for consistency across the site in indicating it's a terminal command, but I'm fine dropping its use here. 

And you're right we likely don't need a shell prompt at all here, since we don't suggest using a command to modify the properties file anyway.

**Review by ericwindmill — APPROVED** [review]
The instructions worked and are easy to follow.

File: src/release/breaking-changes/android-java-gradle-migration-guide.md [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+* Your development machine _might_ contain more
+  than one copy of the Java SDK:
+  * The Android Studio app includes a version of Java,
+    which Flutter uses by default.
+  * If you don't have Android Studio installed,
+    Flutter relies on the version defined by your
+    shell script's `JAVA_HOME` environment variable.
+  * If `JAVA_HOME` isn't defined, Flutter looks
+    for any `java` executable in your path.
+    Once [issue 122609][] lands, the `flutter doctor`
```
Comment (mariamhas) [inline src/release/breaking-changes/android-java-gradle-migration-guide.md:99]: it is only landing for 3.10 right? or will anyone using flutter doctor in previous releases also get this fix so that they can depend on it?

File: src/release/breaking-changes/index.md [hunk 10]
```diff
@@ -30,11 +30,13 @@ release, and listed in alphabetical order:
 
 ### Not yet released to stable
 
+* [Android Java Gradle migration guide][]
```
Comment (mariamhas) [inline src/release/breaking-changes/index.md:33]: hmm this feels weird to be under not yet release to stable, currently there is nothing in the guide that depends on the next stable release for developers to follow it and unblock themselves when using Flamingo.

if we add the flutter analyze --suggestions tool steps into the guide then  it has something that is not yet released to stable

File: src/release/breaking-changes/android-java-gradle-migration-guide.md [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+cd android/gradle/wrapper
+```
+
+**Step 2**: Edit the `gradle-wrapper.properties` file to
+  change the `distributionUrl` field to the new Gradle version:
+
+```properties
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.0.2-all.zip 
+```
+
```
[the remark this answers is not in the document]
Comment (mariamhas) [reply src/release/breaking-changes/android-java-gradle-migration-guide.md:78]: fwiw, i think we should have steps to help developers validate if the error they are seeing is truly due to incompatible versions, considering the PR camille has to improve the error message will land in 3.10 first and hopefully will be pushed into previous stable too.

**Review by mariamhas — DISMISSED** [review]

File: src/release/breaking-changes/android-java-gradle-migration-guide.md (hunk 9 above)
Comment (mariamhas) [reply src/release/breaking-changes/android-java-gradle-migration-guide.md:99]: it being issue 122609

File: src/release/breaking-changes/android-java-gradle-migration-guide.md [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+```sh
+cd android/gradle/wrapper
+```
+
+**Step 2**: Edit the `gradle-wrapper.properties` file to
+  change the `distributionUrl` field to the new Gradle version:
+
+```properties
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.0.2-all.zip
```
Comment (reidbaker) [inline src/release/breaking-changes/android-java-gradle-migration-guide.md:76]: I think we should suggest updating to a version that starts with 7 not the latest gradle.
This is for 2 reasons. In my own updates of our owned examples we have had fewer issues, sometimes non when updating to 7.3.3 (minimum) or  7.6.1 (newest 7.* version released in feb) and run into significantly more issues moving to 8.*. 
Second is that the gradle tooling as a policy has one major version of depreciation cycle before removal. So assuming their gradle code did have something deprecated it is more likely to exist in major version 7 and removed in 8

File: src/release/breaking-changes/android-java-gradle-migration-guide.md [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+```
+
+## Notes
+
+A few notes to be aware of:
+
+* Repeat this step for each affected Android app.
+* This issue can also be experienced by those who
+  don't use Android Studio.
+  If you've manually upgraded the Java SDK to
```
[the remark this answers is not in the document]
Comment (reidbaker) [reply src/release/breaking-changes/android-java-gradle-migration-guide.md:86]: Perhaps: 
This issue can be experienced by anyone who updates Android Studio as flutter defaults to using the version of java bundled with Android Studio.

File: src/release/breaking-changes/android-java-gradle-migration-guide.md [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+  * The Android Studio app includes a version of Java,
+    which Flutter uses by default.
+  * If you don't have Android Studio installed,
+    Flutter relies on the version defined by your
+    shell script's `JAVA_HOME` environment variable.
+  * If `JAVA_HOME` isn't defined, Flutter looks
+    for any `java` executable in your path.
+    Once [issue 122609][] lands, the `flutter doctor`
+    command reports which version of Java is used.
+* If you upgrade Gradle to a release _newer_ than 7.6,
```
Comment (reidbaker) [inline src/release/breaking-changes/android-java-gradle-migration-guide.md:101]: I think this should be 7.6.* or 7.6.2. I think we should do our part to encourage the latest patch versions.

File: src/release/breaking-changes/android-java-gradle-migration-guide.md [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+  	 …
+  	 … 209 more lines of Groovy and Gradle stack trace …
+  	 …
+  	 …]
+  	at java.base/java.lang.Thread.run(Thread.java:833)
+```
+
+This error occurs because Android Studio Flamingo
+updates the Java SDK from 11 to 17, which
+isn't compatible with the version of the Java SDK
```
[the remark this answers is not in the document]
Comment (gmackall) [reply src/release/breaking-changes/android-java-gradle-migration-guide.md:40]: +1, the issue is that the new jdk version (17) isn't supported by the older version of gradle that `flutter create` set up until [around a year ago](https://github.com/flutter/flutter/pull/99723/files#diff-afa4fef23f35d6a99b734a7c1ec54b98750e48714a3026e2f2eccb5468ca488cR30)

File: src/release/breaking-changes/android-java-gradle-migration-guide.md (hunk 7 above)
Comment (gmackall) [reply src/release/breaking-changes/android-java-gradle-migration-guide.md:39]: Maybe something along the lines of "Android Studio Flamingo updates its bundled Java SDK version from..."?

File: src/release/breaking-changes/android-java-gradle-migration-guide.md (hunk 9 above)
Comment (sfshaza2) [reply src/release/breaking-changes/android-java-gradle-migration-guide.md:99]: ??? I don't know what we advise here.

File: src/release/breaking-changes/index.md (hunk 10 above)
Comment (sfshaza2) [reply src/release/breaking-changes/index.md:33]: Yeah. I can just take it out of the index.

File: src/release/breaking-changes/android-java-gradle-migration-guide.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/release/breaking-changes/android-java-gradle-migration-guide.md:78]: If that happens, we can update the error message on this page.  But, yeah, it seems like the error itself is enough to determine if there's an issue.

**Comment by sfshaza2** [comment]
@Updated. Please review again @timsneath @mariamhas @reidbaker

File: src/release/breaking-changes/android-java-gradle-migration-guide.md (hunk 6 above)
Comment (timsneath) [reply src/release/breaking-changes/android-java-gradle-migration-guide.md]: I'd be more explicit then, I think. Rather than saying "as part of this fix", should we simply say "upgrading to Flutter 3.10 does not fix this problem"?

File: src/release/breaking-changes/android-java-gradle-migration-guide.md (hunk 12 above)
Comment (timsneath) [reply src/release/breaking-changes/android-java-gradle-migration-guide.md]: Per https://gradle.org/whats-new/gradle-8/#upgrades, it looks like users should upgrade to 7.6.1 as an intermediary state, even if they're ultimately moving to 8.x -- so I agree that we should make that our target throughout this doc.

File: src/release/breaking-changes/android-java-gradle-migration-guide.md [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+  	 …
+  	 …]
+  	at java.base/java.lang.Thread.run(Thread.java:833)
+```
+
+This error occurs because Android Studio Flamingo
+updates its bundled Java SDK from 11 to 17.
+Flutter uses the version of Java bundled with
+Android Studio to build Android apps.
+Gradle versions [prior to 7.3][] can't run
```
Comment (reidbaker) [inline src/release/breaking-changes/android-java-gradle-migration-guide.md:43]: Is this supposed to have a link?

**Review by reidbaker — APPROVED** [review]
